bundles/org.openhab.ui/web/src/components/widgets/useWidgetExpression.ts

```diff
@@ -0,0 +1,229 @@
+import isYesterday from 'dayjs/plugin/isYesterday'
+import isTomorrow from 'dayjs/plugin/isTomorrow'
+
+import jsepRegex from '@jsep-plugin/regex'
+import jsepArrow from '@jsep-plugin/arrow'
+import jsepObject from '@jsep-plugin/object'
+import jsepTemplate from '@jsep-plugin/template'
+
+import { useUIOptionsStore } from '@/js/stores/useUIOptionsStore'
+import { useStatesStore } from '@/js/stores/useStatesStore'
+import { useUserStore } from '@/js/stores/useUserStore'
+import { i18n } from '@/js/i18n.ts'
+
+import type { WidgetContext, WidgetProps } from './types'
+
+expr.jsep.plugins.register(jsepRegex, jsepArrow, jsepObject, jsepTemplate)
+
+addUnaryOp('@', (itemName: string | undefined): string => {
+  if (itemName === undefined) return '-'
+  const item = useStatesStore().trackedItems[itemName]
+  return (item.displayState !== undefined ? item.displayState : item.state) ?? '-'
+})
+addUnaryOp('@@', (itemName: string | undefined): string => {
+  if (itemName === undefined) return '-'
+  return useStatesStore().trackedItems[itemName]?.state ?? '-'
+})
+addUnaryOp('#', (itemName: string | undefined): number | undefined => {
+  if (itemName === undefined) return undefined
+  return useStatesStore().trackedItems[itemName]?.numericState
+})
+
+dayjs.extend(relativeTime)
+dayjs.extend(calendar)
+dayjs.extend(localizedFormat)
+dayjs.extend(isoWeek)
+dayjs.extend(isToday)
+dayjs.extend(isYesterday)
+dayjs.extend(isTomorrow)
+
+interface ExpressionAstCache {
+  [key: string]: any
+}
+
+interface ScreenInfo {
+  width: number
+  height: number
+  availWidth: number
+  availHeight: number
+  colorDepth: number
+  pixelDepth: number
+  viewAreaWidth: number
+  viewAreaHeight: number
+  appWidth: number
+  appHeight: number
+}
+
+/**
+ * Composable providing the functionality to evaluate widget expressions.
+ *
+ * Widget expression evaluations need access to the current widget context and props.
+ * If they are available at composable instantiation, they can be passed as properties to the composable.
+ * If they, however, aren't available at instantiation (e.g. because they are computed) they can be passed as function parameters later.
+ *
+ * @param properties
+ */
+export function useWidgetExpression (properties: { context?: WidgetContext, props?: WidgetProps } = {}) {
+  // imports
+  const userStore = useUserStore()
+  const uiOptionsStore = useUIOptionsStore()
+
+  // data
+  const exprAst: ExpressionAstCache = {}
+  const viewAreaWidth = ref<number>(0)
+  const viewAreaHeight = ref<number>(0)
+
+  const instance = getCurrentInstance()
+  const global = instance?.appContext.config.globalProperties
+
+  // lifecycle
+  onMounted(() => {
+    nextTick(() => {
+      updateViewAreaDimensions()
+    })
+    window.addEventListener('resize', updateViewAreaDimensions)
+  })
+
+  onUnmounted(() => {
+    window.removeEventListener('resize', updateViewAreaDimensions)
+  })
+
+  // computed
+  const screenWidth = computed(() => window.screen.width)
```

Comment:
I don't believe any that are dependent on window.screen need to be computed/reactive since window.screen itself is not reactive, correct?

Reply:
Correct, will change that.